packages/astro/src/vite-plugin-markdown/index.ts

```diff
@@ -22,14 +27,83 @@ export default function markdown({ config }: AstroPluginOptions): Plugin {
 	}
 
+	// Weird vite behavior: Vite seems to use a fake "index.html" importer when you
+	// have `enforce: pre`. This can probably be removed once the vite issue is fixed.
+	// see: https://github.com/vitejs/vite/issues/5981
+	const fakeRootImporter = fileURLToPath(new URL('index.html', config.projectRoot));
+	function isRootImport(importer: string | undefined) {
+		if (!importer) {
+			return true;
+		}
+		if (importer === fakeRootImporter) {
+			return true;
+		}
+		return false;
+	}
+
 	return {
 		name: 'astro:markdown',
-		enforce: 'pre', // run transforms before other plugins can
+		enforce: 'pre',
+		async resolveId(id, importer, options) {
+			// Resolve virtual modules.
+			if (id.startsWith(resolvedVirtualModuleId)) {
+				return id;
+			}
+			// If the markdown file is imported from another file via ESM, resolve a JS representation
+			// that defers the markdown -> HTML rendering until it is needed. This is especially useful
+			// when fetching and then filtering many markdown files, like with import.meta.glob() or Astro.glob().
+			// Otherwise, resolve directly to the actual component.
+			if (id.endsWith('.md') && !isRootImport(importer)) {
+				const resolvedId = await this.resolve(id, importer, { skipSelf: true, ...options });
+				if (resolvedId) {
+					return resolvedVirtualModuleId + resolvedId.id;
+				}
+			}
+			// Resolve any .md files with the `?content` cache buster. This should only come from
+			// an already-resolved JS module wrapper. Needed to prevent infinite loops in Vite.
+			if (id.endsWith('.md?content')) {
+				const resolvedId = await this.resolve(id, importer, { skipSelf: true, ...options });
+				return resolvedId?.id.replace('?content', '');
+			}
+			return undefined;
+		},
 		async load(id) {
-			if (id.endsWith('.md')) {
-				let source = await fs.promises.readFile(id, 'utf8');
+			// A markdown file has been imported via ESM!
+			// Return the file's JS representation, which includes all Markdown frontmatter data
+			// and a deferred render of the markdown contents.
+			if (id.startsWith(resolvedVirtualModuleId)) {
+				const sitePathname = config.buildOptions.site ? new URL(config.buildOptions.site).pathname : '/';
+				const fileId = id.substring(resolvedVirtualModuleId.length);
+				const fileUrl = fileId.includes('/pages/') ? fileId.replace(/^.*\/pages\//, sitePathname).replace(/(\/index)?\.md$/, '') : undefined;
+				let source = await fs.promises.readFile(fileId, 'utf8');
+				const { data: frontmatter } = matter(source);
+				return {
+					code: `   
+						// Static
+						export const frontmatter = ${JSON.stringify(frontmatter)};
+						export const file = ${JSON.stringify(fileId)};
+						export const url = ${JSON.stringify(fileUrl)};
+						
+						// Deferred
+						export default async function load() {
+							return (await import(${JSON.stringify(fileId + '?content')}));
+						};
+						export function getContent() {
+							return load().then((m) => m.default)
+						};
```

Comment:
Still uses `getContent` instead of `Content`. Is that intentional?

Reply:
I could not, for the life of me, get `Content` to work in this PR. Would love your help if you're free to pair on it sometime in the next few days.

Reply:
Fixed with Nate's help!